java/ql/src/experimental/Security/CWE/CWE-749/UnsafeAndroidAccess.ql

```diff
@@ -0,0 +1,139 @@
+ */
+predicate isJSEnabled(MethodAccess ma) {
+  exists(VarAccess va, MethodAccess jsa |
+    ma.getQualifier() = va and
+    jsa.getQualifier() = va.getVariable().getAnAccess() and
+    jsa.getMethod() instanceof AllowJavaScriptMethod and
+    jsa.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+  )
+}
+
+/**
+ * Load URL method call on the `android.webkit.WebView` object
+ */
+class LoadResourceMethodAccess extends MethodAccess {
+  LoadResourceMethodAccess() {
+    this.getMethod().getDeclaringType() instanceof TypeWebView and
+    (
+      this.getMethod().hasName("loadUrl") or
+      this.getMethod().hasName("postUrl")
+    )
+  }
+}
+
+/**
+ * Method access to external inputs of `android.content.Intent` object
+ */
+class IntentGetExtraMethodAccess extends MethodAccess {
+  IntentGetExtraMethodAccess() {
+    this.getMethod().getName().regexpMatch("get\\w+Extra") and
+    this.getMethod().getDeclaringType() instanceof TypeIntent
+    or
+    this.getMethod().getName().regexpMatch("get\\w+") and
+    this.getQualifier().(MethodAccess).getMethod().hasName("getExtras") and
+    this.getQualifier().(MethodAccess).getMethod().getDeclaringType() instanceof TypeIntent
+  }
+}
+
+/**
+ * Source of loading urls
+ */
+class UntrustedResourceSource extends RemoteFlowSource {
+  UntrustedResourceSource() {
+    exists(MethodAccess ma |
+      ma instanceof IntentGetExtraMethodAccess and
+      this.asExpr().(VarAccess).getVariable().getAnAssignedValue() = ma
+    )
+  }
+
+  override string getSourceType() {
+    result = "user input vulnerable to XSS and sensitive resource disclosure attacks" and
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof AllowUniversalAccessMethod and //High precision match of unsafe resource loading
+      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+    )
+    or
+    result = "user input potentially vulnerable to XSS and sensitive resource disclosure attacks" and
+    not exists(MethodAccess ma |
+      ma.getMethod() instanceof AllowUniversalAccessMethod and //High precision match of unsafe resource loading
+      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+    )
+  }
+}
+
+/**
+ * load externally controlled data from loadUntrustedResource
+ */
+predicate loadUntrustedResource(MethodAccess ma, Expr sink) {
+  ma instanceof LoadResourceMethodAccess and
+  sink = ma.getArgument(0)
+}
+
+/**
+ * Sink of loading urls
+ */
+class UntrustedResourceSink extends DataFlow::ExprNode {
+  UntrustedResourceSink() { loadUntrustedResource(_, this.getExpr()) }
+
+  MethodAccess getMethodAccess() { loadUntrustedResource(result, this.getExpr()) }
+}
+
+class LoadUntrustedResourceConfiguration extends TaintTracking::Configuration {
+  LoadUntrustedResourceConfiguration() { this = "LoadUntrustedResourceConfiguration" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof UntrustedResourceSource }
```

Comment:
Is there a reason to only allow `UntrustedResourceSource` and not any other `RemoteFlowSource`?

Reply:
For Android applications, intent extras are the single main source of <code>RemoteFlowSource</code>, which makes the query concise and clear. 

Also the existing implementation of <code>RemoteFlowSource</code> in the repository still has large room for improvement, which is the reason that I submitted another PR #3812.